app/api/entities/v2_support.ts

```diff
@@ -0,0 +1,124 @@
+import { ObjectId } from 'mongodb';
+
+import templates from 'api/templates';
+import { DefaultEntitiesDataSource } from 'api/entities.v2/database/data_source_defaults';
+import { DenormalizationService } from 'api/relationships.v2/services/service_factories';
+import { search } from 'api/search';
+import { DefaultSettingsDataSource } from 'api/settings.v2/database/data_source_defaults';
+import { MongoTransactionManager } from 'api/common.v2/database/MongoTransactionManager';
+import { getClient } from 'api/common.v2/database/getConnectionForCurrentTenant';
+import { objectIndex } from 'shared/data_utils/objectIndex';
+import { EntitySchema } from 'shared/types/entityType';
+import { MetadataSchema } from 'shared/types/commonTypes';
+import { propertyTypes } from 'shared/propertyTypes';
+import { TemplateSchema } from 'shared/types/templateType';
+import { DefaultRelationshipDataSource } from 'api/relationships.v2/database/data_source_defaults';
+
+const entityTypeCheck = (
+  entity: EntitySchema
+): entity is EntitySchema & {
+  template: ObjectId;
+  sharedId: string;
+  metadata: MetadataSchema;
+} => !!entity.template && !!entity.sharedId && !!entity.metadata;
+
+const performNewRelationshipQueries = async (entities: EntitySchema[]) => {
+  const transactionManager = new MongoTransactionManager(getClient());
+  if (!(await DefaultSettingsDataSource(transactionManager).readNewRelationshipsAllowed())) {
+    return;
+  }
+
+  const templateIdToProperties = objectIndex(
+    await templates.get({ _id: entities.map(e => e.template) }),
+    t => t._id.toString(),
+    t => t.properties.filter((prop: any) => prop.type === 'newRelationship')
+  );
+  const entitiesDataSource = DefaultEntitiesDataSource(transactionManager);
+
+  await Promise.all(
+    entities.filter(entityTypeCheck).map(async entity => {
+      const relProperties = templateIdToProperties[entity.template.toHexString()];
+      if (!relProperties) return;
+
+      const queryedEntity = await entitiesDataSource
+        .getByIds([entity.sharedId], entity.language)
+        .first();
+
+      if (queryedEntity) {
+        relProperties.forEach((relProperty: any) => {
+          entity.metadata[relProperty.name] = queryedEntity.metadata[relProperty.name];
+        });
+      }
+    })
+  );
+  entities.forEach(entity => {
+    entity.obsoleteMetadata = [];
+  });
```

Comment:
Should we try avoiding mutating the entities passed by parameter to the function ?

Reply:
I will take a look. Ideally yes. The inconvenience of immutability is that we don't have a mechanism to properly clone the entity. If we agree that modifying a composed object (think about the metadata and other nested structures) should either mutate it or deeply copy it (nothing in the middle), then we erred on the side of ease for this one :)

Reply:
We'll keep this one as is for now.